src/format/KeePass2Reader.cpp

```diff
@@ -297,35 +368,215 @@ bool KeePass2Reader::readHeaderField()
         setProtectedStreamKey(fieldData);
         break;
 
     case KeePass2::StreamStartBytes:
+        Q_ASSERT(m_version < KeePass2::FILE_VERSION_4);
         setStreamStartBytes(fieldData);
         break;
 
     case KeePass2::InnerRandomStreamID:
+        Q_ASSERT(m_version < KeePass2::FILE_VERSION_4);
         setInnerRandomStreamID(fieldData);
         break;
 
+    case KeePass2::KdfParameters: {
+        QBuffer bufIoDevice(&fieldData);
+        if (!bufIoDevice.open(QIODevice::ReadOnly)) {
+            raiseError("Failed to open buffer for KDF parameters in header");
+            return false;
+        }
+        m_db->setKdfParams(readVariantMap(&bufIoDevice));
+        break;
+    }
+
+    case KeePass2::PublicCustomData:
+        m_db->setPublicCustomData(fieldData);
+        break;
+
     default:
         qWarning("Unknown header field read: id=%d", fieldID);
         break;
     }
 
-    return !m_headerEnd;
+    return true;
+}
+
+bool KeePass2Reader::readInnerHeaderField(QIODevice* device)
+{
+    QByteArray fieldIDArray = device->read(1);
+    if (fieldIDArray.size() != 1) {
+        raiseError("Invalid inner header id size");
+        return false;
+    }
+    KeePass2::InnerHeaderFieldID fieldID = static_cast<KeePass2::InnerHeaderFieldID>(fieldIDArray.at(0));
+
+    bool ok;
+    quint32 fieldLen = Endian::readUInt32(device, KeePass2::BYTEORDER, &ok);
+    if (!ok) {
+        raiseError("Invalid inner header field length");
+        return false;
+    }
+
+    QByteArray fieldData;
+    if (fieldLen != 0) {
+        fieldData = device->read(fieldLen);
+        if (fieldData.size() != fieldLen) {
+            raiseError("Invalid header data length");
+            return false;
+        }
+    }
+
+    switch (fieldID) {
+    case KeePass2::InnerHeaderFieldID::End:
+        return false;
+
+    case KeePass2::InnerHeaderFieldID::InnerRandomStreamID:
+        setInnerRandomStreamID(fieldData);
+        break;
+
+    case KeePass2::InnerHeaderFieldID::InnerRandomStreamKey:
+        setProtectedStreamKey(fieldData);
+        break;
+
+    case KeePass2::InnerHeaderFieldID::Binary:
+        if (fieldLen < 1) {
+            raiseError("Invalid inner header binary size");
+            return false;
+        }
+        m_binaryPool.append(fieldData.mid(1));
+        break;
+
+    default:
+        qWarning("Unknown inner header field read: id=%d", fieldID);
+        break;
+    }
+
+    return true;
+}
+
+QVariantMap KeePass2Reader::readVariantMap(QIODevice* device)
+{
+    bool ok;
+    quint16 version = Endian::readUInt16(device, KeePass2::BYTEORDER, &ok)
+            & KeePass2::VARIANTMAP_CRITICAL_MASK;
+    quint16 maxVersion = KeePass2::VARIANTMAP_VERSION & KeePass2::VARIANTMAP_CRITICAL_MASK;
+    if (!ok || (version > maxVersion)) {
+        raiseError(tr("Unsupported KeePass variant map version."));
+        return QVariantMap();
+    }
+
+    QVariantMap vm;
+    while (true) {
```

Comment:
There is only on break condition (besides several hard returns), so you should use that as loop condition instead of an infinite loop.